### PR TITLE
Add observatory Bonjour service to built iOS Info.plist bundle

### LINKED ISSUE #35707: xcode_backend_test.dart will fail if `flutter precache` hasn't run.
This relies on the real cache containing iOS artifacts for the test case "Xcode backend fails for on unsupported configuration combinations".

### LINKED ISSUE #60634: Running debug on iOS 14 shows "would like to find and connect to devices on your local network" popup
```
2020-06-30 12:52:54.237973-0700 Runner[580:228465] Metal API Validation Enabled
2020-06-30 12:52:54.382810-0700 Runner[580:228698] flutter: Observatory listening on http://127.0.0.1:50735/u2EjnBszARk=/
2020-06-30 12:52:54.676675-0700 Runner[580:228686] fopen failed for data file: errno = 2 (No such file or directory)
2020-06-30 12:52:54.676742-0700 Runner[580:228686] Errors found! Invalidating cache...
2020-06-30 12:52:54.764319-0700 Runner[580:228686] fopen failed for data file: errno = 2 (No such file or directory)
2020-06-30 12:52:54.764380-0700 Runner[580:228686] Errors found! Invalidating cache...
```
flutter_02
```
$ flutter doctor -v
[✓] Flutter (Channel master, 1.20.0-2.0.pre, on macOS 10.16 20A4299v, locale en-US)
    • Flutter version 1.20.0-2.0.pre at /Users/surry/Projects/flutter
    • Framework revision 747fe345f1 (12 days ago), 2020-06-18 08:28:03 -0700
    • Engine revision 965fbbed17
    • Dart version 2.9.0 (build 2.9.0-14.0.dev 2b917f5b6a)

[✗] Android toolchain - develop for Android devices
    ✗ Unable to locate Android SDK.
      Install Android Studio from: https://developer.android.com/studio/index.html
      On first launch it will assist you in installing the Android SDK components.
      (or visit https://flutter.dev/docs/get-started/install/macos#android-setup for detailed instructions).
      If the Android SDK has been installed to a custom location, set ANDROID_SDK_ROOT to that location.
      You may also want to add it to your PATH environment variable.


[✓] Xcode - develop for iOS and macOS (Xcode 11.5)
    • Xcode at /Applications/Xcode-11.app/Contents/Developer
    • Xcode 11.5, Build version 11E608c
    • CocoaPods version 1.9.3

[!] Android Studio (not installed)
    • Android Studio not found; download from https://developer.android.com/studio/index.html
      (or visit https://flutter.dev/docs/get-started/install/macos#android-setup for detailed instructions).

[✓] Connected device (2 available)
    • iPad  • 00008027-001925360106802E • ios        • iOS 14.0
    • macOS • macOS                     • darwin-x64 • macOS 10.16 20A4299v

! Doctor found issues in 2 categories.
```

### PR DESCRIPTION
**Pull request by jmagman**
Reland #64988, which failed post-submit due to https://github.com/flutter/flutter/pull/65124.

## Description

Add `NSBonjourServices` `_dartobservatory._tcp` and `NSLocalNetworkUsageDescription` to debug and profile `Info.plist` in the produced, bundled `.app` (not the user-facing Info.plist created in the template and included in Xcode).

This prompts the "Your app would like to find and connect to devices on your local network" permission popup, which is required to unblock the observatory publisher on iOS 14.

`NSLocalNetworkUsageDescription` is underneath:

Screen Shot 2020-09-02 at 5 36 59 PM

Note: this does not add the keys for add-to-app host apps, so they still have the issue of blocking the observatory publisher.

## Related Issues

Fixes https://github.com/flutter/flutter/issues/60634
Fixes #35707.
Related to conversation at https://github.com/flutter/flutter/issues/63893, but that is for add-to-app.

## Tests

Updated `ios_content_validation_test` to validate Release does not have the keys, but Debug does.
Move `xcode_backend_tests` to the command shard, add new behavior tests.

## Kudos
Thanks to @gaaclarke for suggesting to just add the keys in the generated bundle Info.plist, and not the one that's created in the template.  That avoided a huge migration headache.

## Checklist
- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change
- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*

### CODE REVIEW

File: packages/flutter_tools/bin/xcode_backend.sh [hunk 1]
```diff
@@ -306,6 +320,36 @@ EmbedFlutterFrameworks() {
...
+  # Debug and profile only.
+  if [[ "${build_mode}" == "release" ]]; then
+    return
+  fi
+  local built_products_plist="${BUILT_PRODUCTS_DIR}/${INFOPLIST_PATH}"
+
+  if [[ ! -f "${built_products_plist}" ]]; then
+    EchoError "error: ${INFOPLIST_PATH} does not exist. The Flutter \"Thin Binary\" build phase must run after \"Copy Bundle Resources\"."
+    exit -1
+  fi
```
Comment (jmagman) [inline packages/flutter_tools/bin/xcode_backend.sh:339]: This check is new and was not in https://github.com/flutter/flutter/pull/64988.  If the user rearranged their build scripts manually, error and tell them how to fix it.
Screen Shot 2020-09-02 at 5 16 38 PM

File: packages/flutter_tools/test/integration.shard/xcode_backend_test.dart [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+  group('observatory Bonjour service keys', () {
+    Directory buildDirectory;
+    File infoPlist;
+
+    setUp(() {
+      buildDirectory = globals.fs.systemTempDirectory.createTempSync('flutter_tools_xcode_backend_test.');
+      infoPlist = buildDirectory.childFile('Info.plist');
+    });
+
+    test('fails when the Info.plist is missing', () async {
```
Comment (jmagman) [inline packages/flutter_tools/test/integration.shard/xcode_backend_test.dart:97]: This test is new.

File: packages/flutter_tools/test/integration.shard/xcode_backend_test.dart [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+    final ProcessResult result = await Process.run(
+      xcodeBackendPath,
+      <String>['build'],
+      environment: environment,
+    );
+    expect(result.stderr, startsWith(xcodeBackendErrorHeader));
+    expect(result.exitCode, isNot(0));
+  }
+
+  test('Xcode backend fails with no arguments', () async {
```
Comment (jmagman) [inline packages/flutter_tools/test/integration.shard/xcode_backend_test.dart:55]: This test is new and tests https://github.com/flutter/flutter/pull/65124

**Review by gaaclarke — APPROVED** [review]
LGTM

File: packages/flutter_tools/bin/xcode_backend.sh (hunk 1 above)
Comment (bennibau) [reply packages/flutter_tools/bin/xcode_backend.sh:339]: Well I get this error but the build script "Thin Binary" is after "Copy Bundle Resources"
Bildschirmfoto 2020-09-09 um 11 29 50

Comment (jmagman) [reply packages/flutter_tools/bin/xcode_backend.sh:339]: @bennibau can you file a new issue and @ mention me?  I want to see what's going on in your case.

Please include the output of `flutter doctor -v` and the full exported Xcode run report.
View > Navigator > Reports > Export... 
Screen Shot 2020-09-10 at 1 41 57 PM

Also, a screenshot with those build phases expanded.

Comment (sam-parks) [reply packages/flutter_tools/bin/xcode_backend.sh:339]: I'm also hitting this error with this build script: 

Screen Shot 2020-10-04 at 4 42 50 PM

flutter doctor -v: 

[✓] Flutter (Channel master, 1.23.0-8.0.pre.131, on Mac OS X 10.15.6 19G2021 x86_64, locale en-US)
    • Flutter version 1.23.0-8.0.pre.131 at /Users/samparks/dev/sdks/flutter
    • Framework revision 8485580bb9 (3 hours ago), 2020-10-04 14:47:02 -0400
    • Engine revision a8d7d97bef
    • Dart version 2.11.0 (build 2.11.0-185.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /Users/samparks/Library/Android/sdk
    • Platform android-30, build-tools 30.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.0.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.0.1, Build version 12A7300
    • CocoaPods version 1.9.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 48.1.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.49.3)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.15.0

 
[✓] Connected device (3 available)            
    • iPhone (mobile)  • 00008030-000608880E9B802E • ios            • iOS 13.7
    • Web Server (web) • web-server                • web-javascript • Flutter Tools
    • Chrome (web)     • chrome                    • web-javascript • Google Chrome 85.0.4183.121

• No issues found!

**Pull request merged**